### PR TITLE
Introduce a TablePinningPolicy to control how and when memory is pinned

### PR DESCRIPTION
**Pull request by mrambacher**
This PR adds a TablePinningPolicy to the BlockBasedTableOptions.  This new class is Customizable and can be created from strings.  The class has methods to limit the amount of pin data in the system.

The basic approach is that the unit probes (MayPin) if it believes data is available at the time.  If so, the code does it normal checks and then attempts to pin the new data (PinData).  Once the code is done with the data, UnPinData puts the memory back into the available pool

### CODE REVIEW

File: plugin/speedb/pinning_policy/scoped_pinning_policy_test.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+TEST_F(ScopedPinningPolicyTest, TestLimits) {
+  auto policy = GetScopedPolicy();
+  auto opts = policy->GetOptions<ScopedPinningOptions>();
+  ASSERT_NE(opts, nullptr);
+  auto capacity = opts->capacity;
+  size_t bottom = capacity * opts->bottom_limit / 100;
+  size_t mid = capacity * opts->mid_limit / 100;
+  // size_t used = 0;
```
Comment (udi-speedb) [inline plugin/speedb/pinning_policy/scoped_pinning_policy_test.cc:118]: Please remove

File: plugin/speedb/pinning_policy/scoped_pinning_policy_test.cc [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    }
+    auto scoped = pinning_policy_->CheckedCast<ScopedPinningPolicy>();
+    EXPECT_NE(scoped, nullptr);
+    return scoped;
+  }
+  bool PinData(const TablePinningOptions& tpo, uint8_t type, size_t size,
+               std::vector<std::unique_ptr<PinnedEntry>>& entries) {
+    std::unique_ptr<PinnedEntry> p;
+    if (pinning_policy_->PinData(tpo, type, size, &p)) {
+      entries.emplace_back(std::move(p));
```
Comment (udi-speedb) [inline plugin/speedb/pinning_policy/scoped_pinning_policy_test.cc:45]: Wouldn't hurt to assert that p != nullptr on successfull PinData()

**Comment by udi-speedb** [comment]
@mrambacher - First batch of comments. Haven't finished the review.

**Comment by udi-speedb** [comment]
@mrambacher - In case it wasn't clear. I am waiting for you to address the comments I have given before continuing my review.

File: plugin/speedb/pinning_policy/scoped_pinning_policy.cc [hunk 3]
```diff
@@ -49,78 +68,52 @@ std::string ScopedPinningPolicy::GetId() const {
 
 bool ScopedPinningPolicy::MayPin(const TablePinningOptions& tpo, uint8_t type,
                                  size_t size) const {
-  auto limit = GetUsage();
-  return CheckPin(tpo, type, size, limit);
+  return CheckPin(tpo, type, size, usage_);
```
Comment (udi-speedb) [inline plugin/speedb/pinning_policy/scoped_pinning_policy.cc:71]: What is MayPin()? used for?
And, it should be clear to the users that they MayPin() may return true but actual pinning may still fail.

**Review by udi-speedb — CHANGES_REQUESTED** [review]

**Review by udi-speedb — APPROVED** [review]

File: include/rocksdb/table.h [hunk 4]
```diff
@@ -661,6 +662,9 @@ struct BlockBasedTableOptions {
   //
   // Default: 2
   uint64_t num_file_reads_for_auto_readahead = 2;
+
+  // EXPERIMENTAL
```
[the remark this answers is not in the document]
Comment (mrambacher) [reply include/rocksdb/table.h:660]: > Is it valid to mark it as experimental? What will be the behaviour when not specifiying a pinning policy? Or is experminetal referring to policies that are not the default such as ours?

Experimental means that the interface is may change

File: plugin/speedb/pinning_policy/scoped_pinning_policy.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+// limitations under the License.
+
+#include <map>
+#include <mutex>
+
+#include "table/block_based/table_pinning_policy.h"
+namespace ROCKSDB_NAMESPACE {
+struct TablePinningOptions;
+struct ScopedPinningOptions {
+  static const char* kName() { return "ScopedPinningOptions"; }
```
[the remark this answers is not in the document]
Comment (mrambacher) [reply plugin/speedb/pinning_policy/scoped_pinning_policy.h:23]: > I think this is not in use. Should it remain?

It is used implicitly by RegisterOptions

File: table/block_based/table_pinning_policy.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  TablePinningOptions(int _level, bool _is_bottom, size_t _file_size,
+                      size_t _max_file_size_for_l0_meta_pin)
+      : level(_level),
+        is_bottom(_is_bottom),
+        file_size(_file_size),
+        max_file_size_for_l0_meta_pin(_max_file_size_for_l0_meta_pin) {}
+  int level = -1;
+  bool is_bottom = false;
+  size_t file_size = 0;
+  size_t max_file_size_for_l0_meta_pin = 0;
```
[the remark this answers is not in the document]
Comment (mrambacher) [reply table/block_based/table_pinning_policy.h:32]: > Don't we want to add any additional options that will allow us finer control in our pinning policies? They may be default initialized to some harmless values for rocksdb pinning policy.

What other options would you like to see/use?  These are the ones required by the Speedb and RocksDB pinning policies.  If there are others, I can add them now or they can be added in the future.

File: table/block_based/table_pinning_policy.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+// TablePinningPolicy provides a configurable way to determine when blocks
+// should be pinned in memory for the block based tables.
+//
+// Exceptions MUST NOT propagate out of overridden functions into RocksDB,
+// because RocksDB is not exception-safe. This could cause undefined behavior
+// including data loss, unreported corruption, deadlocks, and more.
+class TablePinningPolicy : public Customizable {
+ public:
+  static const uint8_t kTopLevel = 1;
```
[the remark this answers is not in the document]
Comment (mrambacher) [reply table/block_based/table_pinning_policy.h:54]: > Why aren't these in an enum class of uint8_t size?

They could be an enum.  I thought there would be an arbitrary set of them as we got more of a MemoryManager model and that they may not be known at compile time.

**Review by udi-speedb — APPROVED** [review]

**Comment by erez-speedb** [comment]
build failed 
CC       plugin/speedb/speedb_registry.o
 CC       plugin/speedb/memtable/hash_spd_rep.o
 CC       plugin/speedb/paired_filter/speedb_paired_bloom.o
 CC       plugin/speedb/paired_filter/speedb_paired_bloom_internal.o
 CC       plugin/speedb/pinning_policy/scoped_pinning_policy.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/concurrent_tree.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/keyrange.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/lock_request.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/locktree.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/manager.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/range_buffer.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/treenode.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/txnid_set.o
 CC       utilities/transactions/lock/range/range_tree/lib/locktree/wfg.o
plugin/speedb/pinning_policy/[scoped_pinning_policy.cc](http://scoped_pinning_policy.cc/): In member function ‘virtual bool rocksdb::ScopedPinningPolicy::CheckPin(const rocksdb::TablePinningOptions&, uint8_t, size_t, size_t) const’:
plugin/speedb/pinning_policy/[scoped_pinning_policy.cc:56](http://scoped_pinning_policy.cc:56/):48: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
  56 |   if (tpo.is_bottom && options_.bottom_percent >= 0) {
     |                        ~~~~~~~~~~~~~~~~~~~~~~~~^~~~
plugin/speedb/pinning_policy/[scoped_pinning_policy.cc:60](http://scoped_pinning_policy.cc:60/):52: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
  60 |   } else if (tpo.level > 0 && options_.mid_percent >= 0) {
     |                               ~~~~~~~~~~~~~~~~~~~~~^~~~
 CC       utilities/transactions/lock/range/range_tree/lib/standalone_port.o
 CC       utilities/transactions/lock/range/range_tree/lib/util/dbt.o
 CC       utilities/transactions/lock/range/range_tree/lib/util/memarena.o
 CC       utilities/transactions/lock/range/range_tree/range_tree_lock_manager.o
 CC       utilities/transactions/lock/range/range_tree/range_tree_lock_tracker.o
cc1plus: all warnings being treated as errors
make: *** [Makefile:2418: plugin/speedb/pinning_policy/scoped_pinning_policy.o] Error 1

**Review by udi-speedb — APPROVED** [review]

**Comment by Yuval-Ariel** [comment]
blocked until https://github.com/speedb-io/speedb/issues/573 is fixed @maxb-io FYI

**Review by udi-speedb — APPROVED** [review]

**Review by Yuval-Ariel — APPROVED** [review]

**Comment by Yuval-Ariel** [comment]
@mrambacher plz also add to the HISTORY.md file

**Comment by Yuval-Ariel** [comment]
https://github.com/speedb-io/speedb/actions/runs/5444728254/jobs/9904106132#step:6:571

==15027==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x606000010398 at pc 0x7f21e95a2c2e bp 0x7f21dcdb9f20 sp 0x7f21dcdb9f10
WRITE of size 8 at 0x606000010398 thread T6
    #0 0x7f21e95a2c2d in std::__atomic_base<unsigned long>::operator+=(unsigned long) /usr/include/c++/7/bits/atomic_base.h:312
    #1 0x7f21e95a2c2d in rocksdb::RecordingPinningPolicy::RecordPinned(int, unsigned char, unsigned long, bool) table/block_based/table_pinning_policy.cc:145
    #2 0x7f21e95a2e92 in rocksdb::RecordingPinningPolicy::PinData(rocksdb::TablePinningOptions const&, unsigned char, unsigned long, std::unique_ptr<rocksdb::PinnedEntry, std::default_delete<rocksdb::PinnedEntry> >*) table/block_based/table_pinning_policy.cc:126
    #3 0x7f21e959030e in rocksdb::PartitionIndexReader::Create(rocksdb::BlockBasedTable const*, rocksdb::ReadOptions const&, rocksdb::TablePinningOptions const&, rocksdb::FilePrefetchBuffer*, bool, bool, bool, rocksdb::BlockCacheLookupContext*,

File: db_stress_tool/db_stress_gflags.cc [hunk 8]
```diff
@@ -876,6 +876,8 @@ DEFINE_string(fs_uri, "",
               " with --env_uri."
               " Creates a default environment with the specified filesystem.");
 
+DEFINE_string(pinning_policy, "", "URI for registry TablePinningPolicy");
```
Comment (Yuval-Ariel) [inline db_stress_tool/db_stress_gflags.cc:879]: plz also add the URIs that exist to tools/db_crashtest.py so that the stress test will use this feature

**Review by Yuval-Ariel — CHANGES_REQUESTED** [review]
remove foo file.

**Comment by Yuval-Ariel** [comment]
failed unit tests:
FAILED TESTS (1/11819):
     752 ms: /__w/speedb/speedb/options_settable_test OptionsSettableTest.BlockBasedTableOptionsAllFieldsSettable

**Review by Yuval-Ariel — APPROVED** [review]

**Pull request merged**